src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecc_import_export.c

```diff
@@ -0,0 +1,547 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "pal_ecc_import_export.h"
+#include "pal_utilities.h"
+#include "pal_jni.h"
+#include "pal_eckey.h"
+#include "pal_bignum.h"
+
+int32_t CryptoNative_GetECKeyParameters(
+    const EC_KEY* key,
+    int32_t includePrivate,
+    jobject* qx, int32_t* cbQx,
+    jobject* qy, int32_t* cbQy,
+    jobject* d, int32_t* cbD)
+{
+    assert(qx != NULL);
+    assert(cbQx != NULL);
+    assert(qy != NULL);
+    assert(cbQy != NULL);
+    assert(d != NULL);
+    assert(cbD != NULL);
+
+    JNIEnv* env = GetJNIEnv();
+
+    // Get the public key
+    jobject publicKey = (*env)->CallObjectMethod(env, key->keyPair, g_keyPairGetPublicMethod);
+
+    jobject Q = (*env)->CallObjectMethod(env, publicKey, g_ECPublicKeyGetW);
+
+    (*env)->DeleteLocalRef(env, publicKey);
+
+    jobject xBn = (*env)->CallObjectMethod(env, Q, g_ECPointGetAffineX);
+    jobject yBn = (*env)->CallObjectMethod(env, Q, g_ECPointGetAffineY);
+    
+    (*env)->DeleteLocalRef(env, Q);
+
+    // Success; assign variables
+    *qx = ToGRef(env, xBn); *cbQx = CryptoNative_GetBigNumBytes(xBn); xBn = NULL;
+    *qy = ToGRef(env, yBn); *cbQy = CryptoNative_GetBigNumBytes(yBn); yBn = NULL;
+    if (*cbQx == FAIL || *cbQy == FAIL) goto error;
+
+    if (includePrivate)
+    {
+        jobject privateKey = (*env)->CallObjectMethod(env, key->keyPair, g_keyPairGetPrivateMethod);
+
+        jobject dBn = (*env)->CallObjectMethod(env, privateKey, g_ECPrivateKeyGetS);
+
+        (*env)->DeleteLocalRef(env, privateKey);
+
+        *d = ToGRef(env, dBn);
+        *cbD = CryptoNative_GetBigNumBytes(*d);
+        if (*cbD == FAIL) goto error;
+    }
+    else
+    {
+        if (d)
+            *d = NULL;
+
+        if (cbD)
+            *cbD = 0;
+    }
+
+    return SUCCESS;
+
+error:
+    *cbQx = *cbQy = 0;
+    *qx = *qy = 0;
+    if (d) *d = NULL;
+    if (cbD) *cbD = 0;
+    if (xBn) (*env)->DeleteLocalRef(env, xBn);
+    if (yBn) (*env)->DeleteLocalRef(env, yBn);
+    return FAIL;
+}
+
+int32_t CryptoNative_GetECCurveParameters(
+    const EC_KEY* key,
+    int32_t includePrivate,
+    ECCurveType* curveType,
+    jobject* qx, int32_t* cbQx,
```

Comment:
Might it make sense to create a blittable in/out value type for these large pointer sets? I assume all these functions are for internal usage and not external consumption.

Reply:
My answer to this, and everything else in the shim API, is "do whatever makes sense for this interop".  Just make a new ECDsaAndroid/ECDiffieHellmanAndroid as internal types in System.Security.Cryptography.Algorithms and own the interop for them entirely.  There's no reason to stay aligned with the OpenSSL shim API... especially since I've already started the process of rewriting that one :smile:.